File: Data Structures/src/medium/A Weird Game.cpp
#include <bits/stdc++.h>
#define MAX 100005

using namespace std;

vector <int> F, B;
int tree[MAX];

void validate(string s)
{
    for ( int i = 0; i < (int)s.size(); i++ ) {
        assert(s[i] >= 'a' && s[i] <= 'z');
    }
}

bool chk(string s, string t)
{
    int p = 0;
    for ( int i = 0; i < (int)s.size() && p < (int)t.size(); i++ ) {
        if ( s[i] == t[p] ) p++;
    }
    if ( p == (int)t.size() ) return true;
    return false;
}

void update(int idx)
{
    while ( idx <= MAX - 4) {
        tree[idx] += 1;
        idx += (idx & (-idx));
    }
}

int query(int idx)
{
    int ans = 0;
    while ( idx > 0 ) {
        ans += tree[idx];
        idx -= (idx & (-idx));
    }
    return ans;
}

int main()
{
    int tc, p, idx;
    long long ans;
    string s, t;
    
    cin >> tc;
    assert(tc >= 1 && tc <= 10);
    
    while ( tc-- ) {
        
        cin >> s >> t;
        assert((int)s.size() >= 1 && (int)s.size() <= 100000);
        assert((int)t.size() >= 1 && (int)t.size() <= (int)s.size());
        validate(s);
        validate(t);
        assert(chk(s, t));
        memset(tree, 0, sizeof(tree));
        F.clear(), B.clear();
        
        p = 0;
        for ( int i = 0; i < (int)s.size(); i++ ) {
            if ( p < (int)t.size() && s[i] == t[p] ) p++;
            F.push_back(p);
        }
        
        p = 0;
        
        for ( int i = (int)s.size() - 1; i >= 0; i-- ) {
            if ( p < (int)t.size() && s[i] == t[(int)t.size() - 1 - p] ) p++;
            B.push_back(p);
        }
        
        reverse(B.begin(), B.end());
        
        ans = 0;
        idx = (int)s.size() - 1;
        
        for ( int i = (int)s.size() - 1; i >= 0; i-- ) {
            while ( idx > i + 1 ) {
                if ( B[idx] ) update(B[idx]);
                idx--;
            }
            if ( F[i] == (int)t.size() ) {
                ans += (int)s.size() - i - 1;
            }
            else {
                ans += query((int)t.size()) - query((int)t.size() - F[i] - 1);
            }
        }
        
        for ( int i = 1; i < (int)s.size(); i++ ) {
            ans += (B[i] == (int)t.size());
        }
        
        cout << ans << endl;
    }
    
    
    return 0;
}

File: Dynamic Programming/medium/A Weird Game.java
import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 100005;

    static ArrayList<Integer> F = new ArrayList<>();
    static ArrayList<Integer> B = new ArrayList<>();
    static int[] tree = new int[MAX];

    static void validate(String s) {
        for (char c : s.toCharArray()) {
            if (c < 'a' || c > 'z')
                throw new AssertionError();
        }
    }

    static boolean chk(String s, String t) {
        int p = 0;
        for (int i = 0; i < s.length() && p < t.length(); i++) {
            if (s.charAt(i) == t.charAt(p)) p++;
        }
        return p == t.length();
    }

    // Fenwick update
    static void update(int idx) {
        while (idx <= MAX - 4) {
            tree[idx] += 1;
            idx += idx & -idx;
        }
    }

    // Fenwick query
    static int query(int idx) {
        int ans = 0;
        while (idx > 0) {
            ans += tree[idx];
            idx -= idx & -idx;
        }
        return ans;
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int tc = Integer.parseInt(br.readLine().trim());
        if (tc < 1 || tc > 10) throw new AssertionError();

        StringBuilder out = new StringBuilder();

        while (tc-- > 0) {

            st = new StringTokenizer(br.readLine());
            String s = st.nextToken();
            String t = br.readLine().trim();

            if (s.length() < 1 || s.length() > 100000) throw new AssertionError();
            if (t.length() < 1 || t.length() > s.length()) throw new AssertionError();

            validate(s);
            validate(t);
            if (!chk(s, t)) throw new AssertionError();

            Arrays.fill(tree, 0);
            F.clear();
            B.clear();

            int p = 0;

            // Forward matches
            for (int i = 0; i < s.length(); i++) {
                if (p < t.length() && s.charAt(i) == t.charAt(p)) p++;
                F.add(p);
            }

            p = 0;

            // Backward matches
            for (int i = s.length() - 1; i >= 0; i--) {
                if (p < t.length()
                        && s.charAt(i) == t.charAt(t.length() - 1 - p)) {
                    p++;
                }
                B.add(p);
            }

            Collections.reverse(B);

            long ans = 0;
            int idx = s.length() - 1;

            for (int i = s.length() - 1; i >= 0; i--) {

                while (idx > i + 1) {
                    if (B.get(idx) > 0)
                        update(B.get(idx));
                    idx--;
                }

                if (F.get(i) == t.length()) {
                    ans += s.length() - i - 1;
                } else {
                    ans += query(t.length())
                            - query(t.length() - F.get(i) - 1);
                }
            }

            for (int i = 1; i < s.length(); i++) {
                if (B.get(i) == t.length())
                    ans++;
            }

            out.append(ans).append('\n');
        }

        System.out.print(out);
    }
}
